### PR TITLE
Cache substituted base type

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes https://github.com/dotnet/roslyn/issues/69515.

The regression was introduced by https://github.com/dotnet/roslyn/commit/727569fa3fec7b370f55897adc18ef7001818b2f which adds recursive checking of base types. But those base types of generic types were substituted over and over again. VB doesn't have that problem, it seems to cache them, so I've done a similar thing in C#.

### CODE REVIEW

File: src/Compilers/CSharp/Test/EndToEnd/EndToEndTests.cs [hunk 1]
```diff
@@ -43,7 +44,17 @@ private static void RunInThread(Action action)
             }, 0);
 
             thread.Start();
-            thread.Join();
+            if (timeout is { } t && !Debugger.IsAttached)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/EndToEnd/EndToEndTests.cs:47]: >{ } t [](http://example.com/codeflow?start=27&length=5)

nit: we don't need a second local. `is not null` or `is { }` should suffice.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

File: src/Compilers/CSharp/Test/EndToEnd/EndToEndTests.cs (hunk 1 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/EndToEnd/EndToEndTests.cs:47]: It's s struct, so I think I would then need to write `timeout.Value` or even faster `timeout.GetValueOrDefault`, so `{ } t` seems shorter.

**Comment by jjonescz** [comment]
@dotnet/roslyn-compiler for a second review - would like to backport this for 17.7. Thanks.

File: src/Compilers/CSharp/Portable/Symbols/SubstitutedNamedTypeSymbol.cs [hunk 2]
```diff
@@ -149,8 +149,27 @@ internal sealed override ImmutableArray<NamedTypeSymbol> GetDeclaredInterfaces(C
             return _unbound ? ImmutableArray<NamedTypeSymbol>.Empty : Map.SubstituteNamedTypes(OriginalDefinition.GetDeclaredInterfaces(basesBeingResolved));
         }
 
+        private NamedTypeSymbol _lazyBaseType = ErrorTypeSymbol.UnknownResultType;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/SubstitutedNamedTypeSymbol.cs:152]: >_lazyBaseType [](http://example.com/codeflow?start=32&length=13)

Consider moving field declaration before method declarations.

**Review by cston — APPROVED** [review]

**Review by cston — APPROVED** [review]

**Pull request merged**